File: src/main/java/cz/vhromada/catalog/gui/show/ShowsStatsTableDataModel.java
package cz.vhromada.catalog.gui.show;

import java.util.List;

import cz.vhromada.catalog.common.Time;
import cz.vhromada.catalog.entity.Show;
import cz.vhromada.catalog.facade.ShowFacade;
import cz.vhromada.catalog.gui.common.AbstractStatsTableDataModel;
import cz.vhromada.result.Result;
import cz.vhromada.result.Status;

import org.springframework.util.Assert;

/**
 * A class represents data model for table with stats for shows.
 *
 * @author Vladimir Hromada
 */
@SuppressWarnings("Duplicates")
public class ShowsStatsTableDataModel extends AbstractStatsTableDataModel {

    /**
     * SerialVersionUID
     */
    private static final long serialVersionUID = 1L;

    /**
     * Error message for bad column
     */
    private static final String BAD_COLUMN_ERROR_MESSAGE = "Bad column";

    /**
     * Facade for shows
     */
    private final ShowFacade showFacade;

    /**
     * List of shows
     */
    private List<Show> shows;

    /**
     * Count of seasons from all shows
     */
    private int seasonsCount;

    /**
     * Count of episodes from all shows
     */
    private int episodesCount;

    /**
     * Total length of all shows
     */
    private Time totalLength;

    /**
     * Creates a new instance of ShowsStatsTableDataModel.
     *
     * @param showFacade facade for shows
     * @throws IllegalArgumentException if service is null
     */
    public ShowsStatsTableDataModel(final ShowFacade showFacade) {
        Assert.notNull(showFacade, "Facade for shows mustn't be null.");

        this.showFacade = showFacade;
        update();
    }

    @Override
    public Object getValueAt(final int rowIndex, final int columnIndex) {
        switch (columnIndex) {
            case 0:
                return shows.size();
            case 1:
                return seasonsCount;
            case 2:
                return episodesCount;
            case 3:
                return totalLength.toString();
            default:
                throw new IndexOutOfBoundsException(BAD_COLUMN_ERROR_MESSAGE);
        }
    }

    @Override
    public int getColumnCount() {
        return 4;
    }

    @Override
    public int getRowCount() {
        return 1;
    }

    @Override
    public Class<?> getColumnClass(final int columnIndex) {
        switch (columnIndex) {
            case 0:
                return Integer.class;
            case 1:
                return Integer.class;
            case 2:
                return Integer.class;
            case 3:
                return String.class;
            default:
                throw new IndexOutOfBoundsException(BAD_COLUMN_ERROR_MESSAGE);
        }
    }

    @Override
    public String getColumnName(final int column) {
        switch (column) {
            case 0:
                return "Count of shows";
            case 1:
                return "Count of seasons";
            case 2:
                return "Count of episodes";
            case 3:
                return "Total length";
            default:
                throw new IndexOutOfBoundsException(BAD_COLUMN_ERROR_MESSAGE);
        }
    }

    @Override
    public final void update() {
        final Result<List<Show>> showsResult = showFacade.getAll();
        final Result<Integer> seasonsCountResult = showFacade.getSeasonsCount();
        final Result<Integer> episodesCountResult = showFacade.getEpisodesCount();
        final Result<Time> totalLengthResult = showFacade.getTotalLength();

        final Result<Void> result = new Result<>();
        result.addEvents(showsResult.getEvents());
        result.addEvents(seasonsCountResult.getEvents());
        result.addEvents(episodesCountResult.getEvents());
        result.addEvents(totalLengthResult.getEvents());

        if (Status.OK == result.getStatus()) {
            shows = showsResult.getData();
            seasonsCount = seasonsCountResult.getData();
            episodesCount = episodesCountResult.getData();
            totalLength = totalLengthResult.getData();
        } else {
            throw new IllegalArgumentException("Can't get data. " + result);
        }
    }

}

File: src/main/kotlin/cz/vhromada/catalog/gui/show/ShowsStatsTableDataModel.kt
package cz.vhromada.catalog.gui.show

import cz.vhromada.catalog.entity.Show
import cz.vhromada.catalog.facade.ShowFacade
import cz.vhromada.catalog.gui.common.AbstractStatsTableDataModel
import cz.vhromada.common.Time
import cz.vhromada.common.result.Result
import cz.vhromada.common.result.Status

/**
 * A class represents data model for table with stats for shows.
 *
 * @author Vladimir Hromada
 */
class ShowsStatsTableDataModel(private val showFacade: ShowFacade) : AbstractStatsTableDataModel() {

    /**
     * List of shows
     */
    private lateinit var shows: List<Show>

    /**
     * Count of seasons from all shows
     */
    private var seasonsCount: Int = 0

    /**
     * Count of episodes from all shows
     */
    private var episodesCount: Int = 0

    /**
     * Total length of all shows
     */
    private lateinit var totalLength: Time

    init {
        update()
    }

    override fun getValueAt(rowIndex: Int, columnIndex: Int): Any {
        return when (columnIndex) {
            0 -> shows.size
            1 -> seasonsCount
            2 -> episodesCount
            3 -> totalLength.toString()
            else -> throw IndexOutOfBoundsException(BAD_COLUMN_ERROR_MESSAGE)
        }
    }

    override fun getColumnCount(): Int {
        return 4
    }

    override fun getRowCount(): Int {
        return 1
    }

    override fun getColumnClass(columnIndex: Int): Class<*> {
        return when (columnIndex) {
            0 -> Int::class.java
            1 -> Int::class.java
            2 -> Int::class.java
            3 -> String::class.java
            else -> throw IndexOutOfBoundsException(BAD_COLUMN_ERROR_MESSAGE)
        }
    }

    override fun getColumnName(column: Int): String {
        return when (column) {
            0 -> "Count of shows"
            1 -> "Count of seasons"
            2 -> "Count of episodes"
            3 -> "Total length"
            else -> throw IndexOutOfBoundsException(BAD_COLUMN_ERROR_MESSAGE)
        }
    }

    @Suppress("DuplicatedCode")
    override fun update() {
        val showsResult = showFacade.getAll()
        val seasonsCountResult = showFacade.getSeasonsCount()
        val episodesCountResult = showFacade.getEpisodesCount()
        val totalLengthResult = showFacade.getTotalLength()

        val result = Result<Void>()
        result.addEvents(showsResult.events())
        result.addEvents(seasonsCountResult.events())
        result.addEvents(episodesCountResult.events())
        result.addEvents(totalLengthResult.events())

        if (Status.OK == result.status) {
            shows = showsResult.data!!
            seasonsCount = seasonsCountResult.data!!
            episodesCount = episodesCountResult.data!!
            totalLength = totalLengthResult.data!!
        } else {
            throw IllegalArgumentException("Can't get data. $result")
        }
    }

    companion object {

        /**
         * Error message for bad column
         */
        private const val BAD_COLUMN_ERROR_MESSAGE = "Bad column"

    }

}
